packages/aws-cdk/lib/api/cloudformation-deployments.ts

```diff
@@ -366,6 +360,115 @@ export class CloudFormationDeployments {
+
+      if (parentStackName && !listStackResourcesByStackName[parentStackName]) {
+        listStackResourcesByStackName[parentStackName] = new LazyListStackResources(sdk, parentStackName);
+      }
+
+      const assetPath = nestedStackResource.Metadata['aws:asset:path'];
+      const nestedStackTemplates = await this.getNestedStackTemplates(
+        rootStackArtifact, assetPath, nestedStackLogicalId, parentStackName, listStackResourcesByStackName,
+      );
+
+      generatedParentTemplate.Resources[nestedStackLogicalId] = nestedStackTemplates.generatedNestedTemplate;
+      generatedParentTemplate.Resources[nestedStackLogicalId].Type = 'AWS::CloudFormation::Stack';
+
+      if (deployedParentTemplate.Resources) {
+        deployedParentTemplate.Resources[nestedStackLogicalId] = nestedStackTemplates.deployedNestedStackTemplate;
+        deployedParentTemplate.Resources[nestedStackLogicalId].Type = 'AWS::CloudFormation::Stack';
+      }
+
+      await this.replaceNestedStacksInParentTemplate(
+        rootStackArtifact, generatedParentTemplate.Resources[nestedStackLogicalId],
+        (deployedParentTemplate.Resources ?? {})[nestedStackLogicalId] ?? {},
+        nestedStackTemplates.nestedStackName, listStackResourcesByStackName, sdk,
+      );
+    }
+  }
+
+  private async getNestedStackTemplates(
+    rootStackArtifact: cxapi.CloudFormationStackArtifact, nestedTemplateAssetPath: string, nestedStackLogicalId: string,
+    parentStackName: string | undefined, listStackResourcesByStackName: { [key:string]: ListStackResources },
+  ): Promise<NestedStackTemplates> {
+    const nestedTemplatePath = path.join(rootStackArtifact.assembly.directory, nestedTemplateAssetPath);
+    // CFN generates the nested stack name in the form `ParentStackName-NestedStackLogicalID-SomeHashWeCan'tCompute,
+    // so we get the ARN and manually extract the name.
+    let evaluateCfnTemplate = undefined;
+    if (parentStackName) {
+      evaluateCfnTemplate = new EvaluateCloudFormationTemplate({
+        stackArtifact: rootStackArtifact,
+        parameters: {},
+        account: '',
+        region: '',
+        partition: '',
+        urlSuffix: () => '',
+        listStackResources: listStackResourcesByStackName[parentStackName],
+      });
+    }
+    const nestedStackArn = await evaluateCfnTemplate?.findPhysicalNameFor(nestedStackLogicalId);
+    const nestedStackName = nestedStackArn?.slice(nestedStackArn.indexOf('/') + 1, nestedStackArn.lastIndexOf('/'));
+
+    return {
+      generatedNestedTemplate: JSON.parse(fs.readFileSync(nestedTemplatePath, 'utf-8')),
+      deployedNestedStackTemplate: nestedStackName
+        ? await this.readCurrentNestedTemplate(rootStackArtifact, nestedStackName) : { Resources: {} },
+      nestedStackName,
+    };
+  }
+
+  private isNestedStackResource(stackResource: any): boolean {
+    return stackResource.Type === 'AWS::CloudFormation::Stack' && stackResource.Metadata && stackResource.Metadata['aws:asset:path'];
```

Comment:
How about we add some type safety here, instead of constantly working on `any`?

You know a lot about `stackResource` here, what if you introduced a type for it? Something like

```ts
interface NestedStackResource {
  readonly Metadata: { 'aws:asset:path': string };
}
```

And then in this method, you would do:

```ts
  private isCdkManagedNestedStack(resource: any): resource is NestedStackResource {
    // ...
  }
```